### PR TITLE
new fields and save_saml_account- action for sso

### PR DESCRIPTION
**Pull request by r-peschke**
First commit with models.yml changes and initial-data.json and example-data.json
There will be no migration, if needed take the defaults from models.yml or initial-data.json

### CODE REVIEW

File: openslides_backend/action/actions/user/save_saml_account.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            yield self.build_event(EventType.Create, fqid, instance)
+        else:
+            fields = {
+                k: v
+                for k, v in instance.items()
+                if k != "id" and not k.startswith("meta_")
+            }
+            if not fields:
+                return []
+            yield self.build_event(EventType.Update, fqid, fields)
```
Comment (jsangmeister) [inline openslides_backend/action/actions/user/save_saml_account.py:131]: You can inherit from both `CreateAction` and `UpdateAction` and call the respective `super(SuperClass).create_events()` methods here to reduce code duplication

**Review by jsangmeister — CHANGES_REQUESTED** [review]

File: global/data/example-data.json [hunk 2]
```diff
@@ -32,7 +32,20 @@
...
+                "username": "saml_id",
+                "title": "title",
+                "firstName": "first_name",
+                "lastName": "last_name",
+                "email": "email",
+                "gender": "gender",
+                "pronomen": "pronoun",
+                "is_active": "is_active",
+                "is_person": "is_physical_person"
+              }
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply global/data/example-data.json:48]: This is the only part in the software to give an example and to document what and how to configure the mapping of attributes. That's the reason why I put it here. It could be hidden in the client, but if it will be enabled these base settings should be there.

File: global/data/initial-data.json [hunk 3]
```diff
@@ -18,7 +18,20 @@
...
+                "username": "saml_id",
+                "title": "title",
+                "firstName": "first_name",
+                "lastName": "last_name",
+                "email": "email",
+                "gender": "gender",
+                "pronomen": "pronoun",
+                "is_active": "is_active",
+                "is_person": "is_physical_person"
+              }
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply global/data/initial-data.json:34]: see above

File: global/data/example-data.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsangmeister) [reply global/data/example-data.json:48]: It does not need to be documented in the software. We should add a hint in the client when enabling SSO as well as document this behaviour in some README, but since we will be the ones to enter these details for almost all instances, it does not need to be present in the datastore of all instances when 99% of them don't use the feature.

File: global/data/initial-data.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (jsangmeister) [reply global/data/initial-data.json:34]: As stated above, the initial data should not contain this data. It is meant as the absolute barebone for OpenSlides to work and SSO configuration is not part of that.

File: openslides_backend/action/actions/user/save_saml_account.py (hunk 1 above)
Comment (jsangmeister) [reply openslides_backend/action/actions/user/save_saml_account.py:131]: or, alternatively, move the code to a mixin and use it here and in `UpdateAction`

File: global/meta/models.yml [hunk 4]
```diff
@@ -123,6 +123,19 @@ organization:
       - cs
     required: true
 
+  # SSO settings
+  sso_enabled:
+    type: boolean
+    default: False
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply global/meta/models.yml:129]: done

File: global/meta/models.yml [hunk 5]
```diff
@@ -123,6 +123,19 @@ organization:
...
+    restriction_mode: A
+  login_button_text:
+    type: string
+    default: SAML login
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply global/meta/models.yml:133]: This is the default and a lot of users will stay fine with it without changing.

File: global/meta/models.yml [hunk 6]
```diff
@@ -123,6 +123,19 @@ organization:
...
+    restriction_mode: A
+  save_attr_config:
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply global/meta/models.yml:135]: renamed to `saml_attr_mapping`

File: global/meta/models.yml [hunk 7]
```diff
@@ -123,6 +123,19 @@ organization:
       - cs
     required: true
 
+  # SSO settings
+  sso_enabled:
+    type: boolean
+    default: False
+    restriction_mode: A
+  login_button_text:
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply global/meta/models.yml:131]: renamed to `sso_login_button_text`

File: openslides_backend/action/actions/organization/update.py [hunk 8]
```diff
@@ -22,72 +24,78 @@ class OrganizationUpdate(
...
+        if "save_attr_config" in instance:
+            save_attr_config: Optional[Dict] = instance.get("save_attr_config")
+            if isinstance(save_attr_config, str):
+                try:
+                    save_attr_config = json.loads(save_attr_config)
+                    instance["save_attr_config"] = save_attr_config
+                except json.JSONDecodeError as e:
+                    raise ActionException(
+                        f"save_attr_config must be a valid configuration dictionary for SSO: {str(e)}"
+                    )
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply openslides_backend/action/actions/organization/update.py:88]: removed

File: openslides_backend/action/actions/organization/update.py [hunk 9]
```diff
@@ -22,72 +24,78 @@ class OrganizationUpdate(
...
         ) and not has_organization_management_level(
             self.datastore,
             self.user_id,
             OrganizationManagementLevel.SUPERADMIN,
         ):
             raise MissingPermission(OrganizationManagementLevel.SUPERADMIN)
 
+    def validate_instance(self, instance: Dict[str, Any]) -> None:
+        if "save_attr_config" in instance:
+            save_attr_config: Optional[Dict] = instance.get("save_attr_config")
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply openslides_backend/action/actions/organization/update.py:80]: done

File: openslides_backend/action/actions/organization/update.py [hunk 10]
```diff
@@ -22,72 +24,78 @@ class OrganizationUpdate(
...
+                    save_attr_config = json.loads(save_attr_config)
+                    instance["save_attr_config"] = save_attr_config
+                except json.JSONDecodeError as e:
+                    raise ActionException(
+                        f"save_attr_config must be a valid configuration dictionary for SSO: {str(e)}"
+                    )
+            if not isinstance(instance.get("save_attr_config"), dict):
+                raise ActionException(
+                    "save_attr_config must be a valid configuration dictionary for SSO"
+                )
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply openslides_backend/action/actions/organization/update.py:92]: done

File: openslides_backend/action/actions/organization/update.py [hunk 11]
```diff
@@ -22,72 +24,78 @@ class OrganizationUpdate(
...
+            if "saml_id" not in cast(Dict[Any, Any], save_attr_config).values():
+                raise ActionException(
+                    "save_attr_config must contain the OpenSlides field 'saml_id'"
+                )
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply openslides_backend/action/actions/organization/update.py:96]: This is not the payload field, but a value in the object

File: openslides_backend/action/actions/user/save_saml_account.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+                "SingleSignOn field attributes are not configured in OpenSlides"
+            )
+        new_instance = {
+            attr_config[key]: value
+            for key, value in instance.items()
+            if key in attr_config
+            and attr_config[key] in self.schema["properties"].keys()
+        }
+        instance.clear()
+        instance.update(new_instance)
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply openslides_backend/action/actions/user/save_saml_account.py:68]: This was, because instance is only passed by ref, but not returned.
Extended the action framework to give a chance to modify the instance before validating the schema.

File: openslides_backend/action/actions/user/save_saml_account.py [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+            raise ActionException(
+                "There is no field of user's data mapped to the required 'saml_id'"
+            )
+        super().validate_instance(instance)
+
+    def prefetch(self, action_data: ActionData) -> None:
+        if len(cast(Sized, action_data)) != 1:
+            raise ActionException(
+                "The save_saml_account action accepts only one user instance!"
+            )
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply openslides_backend/action/actions/user/save_saml_account.py:79]: done

File: openslides_backend/action/actions/user/save_saml_account.py [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+    def prepare_action_data(self, action_data: ActionData) -> ActionData:
+        for instance in action_data:
+            users = self.datastore.filter(
+                "user", FilterOperator("saml_id", "=", instance["saml_id"]), ["id"]
+            )
+            if len(users) == 1:
+                instance["id"] = next(iter(users.values()))["id"]
+            elif len(users) == 0:
+                new_id = self.datastore.reserve_ids(self.model.collection, 1)[0]
+                instance["id"] = new_id
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply openslides_backend/action/actions/user/save_saml_account.py:90]: done

File: openslides_backend/action/actions/user/save_saml_account.py [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+                instance["can_change_own_password"] = False
+                instance["organization_id"] = ONE_ORGANIZATION_ID
+                instance["username"] = self.generate_usernames([instance["saml_id"]])[0]
+                instance["meta_new"] = True
+            else:
+                ActionException(
+                    f"More than one existing user found in database with saml_id {instance['aml_id']}"
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply openslides_backend/action/actions/user/save_saml_account.py:97]: done

File: openslides_backend/action/generics/update.py [hunk 16]
```diff
@@ -25,9 +25,13 @@ def create_events(self, instance: Dict[str, Any]) -> Iterable[Event]:
...
+        fields = UpdateAction.create_events_for_reuse(instance)
         if not fields:
             return []
         yield self.build_event(EventType.Update, fqid, fields)
+
+    @staticmethod
+    def create_events_for_reuse(instance: Dict[str, Any]) -> Dict[str, Any]:
+        return {
+            k: v for k, v in instance.items() if k != "id" and not k.startswith("meta_")
+        }
```
Comment (jsangmeister) [inline openslides_backend/action/generics/update.py:37]: Why not just use a global method defined somewhere accessible from both locations? That would be more pythonic and in line with the rest of the backend.

**Review by jsangmeister — CHANGES_REQUESTED** [review]

File: global/data/example-data.json [hunk 17]
```diff
@@ -32,7 +32,20 @@
...
+            "saml_enabled": false,
+            "saml_login_button_text": "SAML Login",
+            "saml_attr_mapping": {
+                "username": "saml_id",
+                "title": "title",
+                "firstName": "first_name",
+                "lastName": "last_name",
+                "email": "email",
+                "gender": "gender",
+                "pronomen": "pronoun",
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply global/data/example-data.json:45]: done

File: global/data/initial-data.json [hunk 18]
```diff
@@ -18,7 +18,20 @@
...
+            "saml_enabled": false,
+            "saml_login_button_text": "SAML Login",
+            "saml_attr_mapping": {
+                "username": "saml_id",
+                "title": "title",
+                "firstName": "first_name",
+                "lastName": "last_name",
+                "email": "email",
+                "gender": "gender",
+                "pronomen": "pronoun",
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply global/data/initial-data.json:31]: done

File: global/meta/models.yml [hunk 19]
```diff
@@ -214,6 +226,7 @@ user:
   saml_id:
     type: string
     restriction_mode: A
+    description: unique-key from IdP for SignleSignOn
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply global/meta/models.yml:229]: done

File: openslides_backend/action/generics/update.py (hunk 16 above)
Comment (r-peschke) [reply openslides_backend/action/generics/update.py:37]: This is global and usable from everywhere, but it is not "somewhere", it is on the place, where it is used in general.

File: openslides_backend/action/actions/user/save_saml_account.py [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+        """
+        fqid = fqid_from_collection_and_id(self.model.collection, instance["id"])
+        if "meta_new" in instance:
+            del instance["meta_new"]
+            yield self.build_event(EventType.Create, fqid, instance)
+        else:
+            fields = UpdateAction.create_events_for_reuse(instance)
+            if not fields:
+                return []
+            yield self.build_event(EventType.Update, fqid, fields)
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply openslides_backend/action/actions/user/save_saml_account.py:136]: No, the event type is different

File: openslides_backend/action/generics/update.py (hunk 16 above)
Comment (jsangmeister) [reply openslides_backend/action/generics/update.py:37]: Mmh, I'm not a fan. But if you insist, okay. But please use a more appropriate name like `create_update_events`.

File: openslides_backend/action/actions/user/save_saml_account.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (jsangmeister) [reply openslides_backend/action/actions/user/save_saml_account.py:136]: How? The type is `Update` in both cases. The lines are exactly the same

File: openslides_backend/action/actions/user/save_saml_account.py [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+                if value == "saml_id"
+            },
+            additional_optional_fields={
+                key: cast(Field, getattr(self.model, value, {})).get_payload_schema()
+                for key, value in self.saml_attr_mapping.items()
+                if value != "saml_id" and value in self.allowed_user_fields
+            },
+            title="create saml account schema",
+        )
+        self.schema["additionalProperties"] = True
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply openslides_backend/action/actions/user/save_saml_account.py:81]: done

File: openslides_backend/action/actions/organization/update.py [hunk 22]
```diff
@@ -22,72 +22,67 @@ class OrganizationUpdate(
...
             self.datastore,
             self.user_id,
             OrganizationManagementLevel.SUPERADMIN,
         ):
             raise MissingPermission(OrganizationManagementLevel.SUPERADMIN)
 
+    def validate_instance(self, instance: Dict[str, Any]) -> None:
+        super().validate_instance(instance)
+        if "saml_attr_mapping" in instance:
+            saml_attr_mapping: Optional[Dict] = instance["saml_attr_mapping"]
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply openslides_backend/action/actions/organization/update.py:80]: whole method removed, because not necessary with new schema check

File: openslides_backend/action/actions/organization/update.py [hunk 23]
```diff
@@ -22,72 +22,67 @@ class OrganizationUpdate(
...
+            if "saml_id" not in cast(Dict[Any, Any], saml_attr_mapping).values():
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply openslides_backend/action/actions/organization/update.py:81]: see remark above

File: openslides_backend/action/actions/user/save_saml_account.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (r-peschke) [reply openslides_backend/action/actions/user/save_saml_account.py:136]: With the check for changed values at all, it is different. We should think about using this everywhere

File: openslides_backend/action/actions/organization/update.py [hunk 24]
```diff
@@ -22,65 +23,61 @@ class OrganizationUpdate(
...
-            "users_email_subject",
-            "users_email_body",
-            "default_language",
-        ]
+        optional_properties=group_A_fields + group_B_fields,
+        additional_optional_fields={
+            "saml_attr_mapping": {
+                "type": "object",
+                "properties": {
+                    field: {"type": ["string", "null"], "maxLength": 256}
```
Comment (jsangmeister) [inline openslides_backend/action/actions/organization/update.py:59]: You could use `optional_str_schema` and insert the `maxLength` into it. I think a `minLength` would be good

File: openslides_backend/action/actions/user/save_saml_account.py [hunk 25]
```diff
@@ -0,0 +1,158 @@
...
+            "$schema": schema_version,
+            "title": "create saml account schema",
+            "type": "object",
+            "properties": {
+                **additional_required_fields,
+                **additional_optional_fields,
+            },
+            "required": list(additional_required_fields.keys()),
+            "additionalProperties": True,
+        }
```
Comment (jsangmeister) [inline openslides_backend/action/actions/user/save_saml_account.py:89]: Multiple remarks:
- You do not need to split the fields artificially into required and optional, just to merge them together again afterwards
- You can use `model.get_field()` instead of `getattr()` to get the field
- You do not need to check for the `allowed_user_fields` since it is guaranteed from `organization.update` that these are correct
- I would use more verbose names instead of `key` and `value`
```suggestion
        schema = {
            "$schema": schema_version,
            "title": "create saml account schema",
            "type": "object",
            "properties": {
                payload_field: self.model.get_field(model_field).get_payload_schema()
                for model_field, payload_field in self.saml_attr_mapping.items()
            },
            "required": self.saml_attr_mapping["saml_id"],
            "additionalProperties": True,
        }
```

File: openslides_backend/action/actions/user/save_saml_account.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (jsangmeister) [reply openslides_backend/action/actions/user/save_saml_account.py:153]: Agreed, that has been on the backlog (at least my mental one) for a long time

**Review by jsangmeister — CHANGES_REQUESTED** [review]

File: openslides_backend/action/actions/user/save_saml_account.py [hunk 26]
```diff
@@ -0,0 +1,158 @@
...
+        Handles create and update
+        """
+        fqid = fqid_from_collection_and_id(self.model.collection, instance["id"])
+        if "meta_new" in instance:
+            del instance["meta_new"]
+            yield self.build_event(EventType.Create, fqid, instance)
+        else:
+            fields = UpdateAction.create_update_events(instance)
+            if not fields:
+                return []
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply openslides_backend/action/actions/user/save_saml_account.py]: done

File: openslides_backend/action/actions/organization/update.py (hunk 24 above)
Comment (r-peschke) [reply openslides_backend/action/actions/organization/update.py:62]: done

File: openslides_backend/action/actions/user/save_saml_account.py (hunk 25 above)
Comment (r-peschke) [reply openslides_backend/action/actions/user/save_saml_account.py]: > Multiple remarks:
> 
>     * You do not need to split the fields artificially into required and optional, just to merge them together again afterwards
done
> 
>     * You can use `model.get_field()` instead of `getattr()` to get the field
> 
done
>     * You do not need to check for the `allowed_user_fields` since it is guaranteed from `organization.update` that these are correct
> 
no, because this could be changed

>     * I would use more verbose names instead of `key` and `value`

done

File: openslides_backend/action/actions/user/save_saml_account.py [hunk 27]
```diff
@@ -99,17 +99,20 @@ def prepare_action_data(self, action_data: ActionData) -> ActionData:
...
+                    tx[0]
+                    if isinstance((tx := instance_old[payload_field]), list) and len(tx)
+                    else tx
+                )
+                if value not in (None, []):
+                    instance[model_field] = value
         users = self.datastore.filter(
             "user",
-            FilterOperator("saml_id", "=", instance["saml_id"]),
+            FilterOperator("saml_id", "=", instance.get("saml_id", "")),
```
Comment (jsangmeister) [inline openslides_backend/action/actions/user/save_saml_account.py:115]: Why did you change this? It's impossible to arrive here without the `saml_id` set

File: openslides_backend/action/actions/user/save_saml_account.py [hunk 28]
```diff
@@ -139,18 +144,13 @@ def create_events(self, instance: Dict[str, Any]) -> Iterable[Event]:
...
-            if not fields:
-                return []
-            fields = {k: v for k, v in instance.items() if v != self.user.get(k)}
-            if not fields:
-                return []
-            yield self.build_event(EventType.Update, fqid, fields)
+            fields = {
+                k: v for k, v in instance.items() if k == "id" or v != self.user.get(k)
+            }
+            yield from UpdateAction.create_events(cast(UpdateAction, self), fields)
```
Comment (jsangmeister) [inline openslides_backend/action/actions/user/save_saml_account.py:153]: Now you're emulating a static method without it actually being one, this is really hacky. What I had in mind was that you just copy the 
```
if not fields:
    return []
yield self.build_event(EventType.Update, fqid, fields)
```
which could by the way rewritten to
```
if fields:
    yield self.build_event(EventType.Update, fqid, fields)
```
from the method in the update action to here. Alternatively, if you want to keep the code deduplication, you have to reintroduce the static method or use one of the solutions we discussed (like inheritance, mixin, etc.). But this is not the way to go...

File: openslides_backend/action/actions/user/save_saml_account.py [hunk 29]
```diff
@@ -66,25 +64,27 @@ def validate_instance(self, instance: Dict[str, Any]) -> None:
...
-        }
         self.schema = {
             "$schema": schema_version,
             "title": "create saml account schema",
             "type": "object",
             "properties": {
-                **additional_required_fields,
-                **additional_optional_fields,
+                payload_field: {
+                    "oneOf": [
```
Comment (jsangmeister) [inline openslides_backend/action/actions/user/save_saml_account.py:73]: `oneOf` will actually lead to such bad error message which you always criticize ;) Why do we need the option to give arrays if we only use the first value anyway?

Also, this will lead to problems once we need the option to fill array fields in the user (e.g. groups).

File: openslides_backend/action/actions/user/save_saml_account.py [hunk 30]
```diff
@@ -66,25 +64,27 @@ def validate_instance(self, instance: Dict[str, Any]) -> None:
...
+                        (
+                            type_def := self.model.get_field(
+                                model_field
+                            ).get_payload_schema()
+                        ),
+                        {"type": "array", "items": type_def}
+                        if model_field != "saml_id"
+                        else {"type": "array", "items": type_def, "minItems": 1},
```
Comment (jsangmeister) [inline openslides_backend/action/actions/user/save_saml_account.py:81]:
```suggestion
                        {"type": "array", "items": type_def, "minItems": int(model_field == "saml_id")},
```
or, if that's too code-golfy:
```suggestion
                        {"type": "array", "items": type_def, "minItems": 1 if model_field == "saml_id" else 0},
```

**Review by jsangmeister — CHANGES_REQUESTED** [review]

File: openslides_backend/action/actions/user/save_saml_account.py (hunk 29 above)
Comment (jsangmeister) [reply openslides_backend/action/actions/user/save_saml_account.py:73]: To get rid of the `oneOf`, you could merge the two schemas by using i.e. `type: ["string", "array"]`. But I don't know if the code will get much better by doing this...

Comment (r-peschke) [reply openslides_backend/action/actions/user/save_saml_account.py:73]: We had to do, it's customer format

File: openslides_backend/action/actions/user/save_saml_account.py (hunk 28 above)
Comment (r-peschke) [reply openslides_backend/action/actions/user/save_saml_account.py:155]: Now I'm calling the super method's of the correct base classes, the update and the create one's. There is no hack at all!

File: openslides_backend/action/actions/user/save_saml_account.py (hunk 27 above)
Comment (r-peschke) [reply openslides_backend/action/actions/user/save_saml_account.py:117]: I did it, because I run into the error

**Review by jsangmeister — APPROVED** [review]
minor change
